# Pruebas de la búsqueda de secuencias con los ÁNGULOS de los esqueletos

### Autora: Lucía Núñez Calvo

#### Fecha: 4 de Mayo del 2022


En este cuaderno se van a mostrar una serie de pruebas de como el algoritmo encuentra la secuencia correspondiente a un ejercicio dentro de una secuencia de mayor tamaño obtenida a partir de un video que realiza varios ejercicios seguidos.

Esto se realizará una vez se hayan obtenido los <b> ángulos </b> de todos los frames, tanto de los ejercicios individuales como del vídeo que realiza éstos ejercicios de manera ininterrumpida. Para ello se han planteado varias soluciones:
- Solución 1: Buscar toda la secuencia en el video completo
- Solución 2: Buscar el inicio y el final
- Solución 3: Buscar la parte "central" del ejercicio

### Planteamiento del problema

Con las posiciones localizadas en las que comienza y finalizan los ejercicios dentro de la secuencia larga se procederá de la siguiente manera. Sabemos que el vídeo en el que se realizan todos los ejercicios tiene "x" frames y sabemos en que segundo del vídeo se empieza y acaba un ejercicio (esto lo observamos al reproducir el vídeo, no es algo que haya obtenido del programa). Con esta información localizaremos en que puntos empiezan y acaban cada uno de los ejercicios y los compararemos con los resultados obtenidos en el programa

Según se ha observado en el vídeo los ejercicios comienzan en los siguientes instantes de tiempo:
- Ejercicio 1: 0 - 0:32 
- Ejercicio 2: 0:32 - 0:51 
- Ejercicio 3: 0:51 - 1:12 
- Ejercicio 4: 1:12 - 1:37

Por lo tanto lo que se tendrá que hacer es dividir el número de frames obtenidos entre el tiempo del vídeo y de esta manera se podrá hacer una aproximación de los frames que ha obtenido en cada segundo.
- El número de frames obtenidos en el vídeo largo han sido <b> 2919 frames </b>
- Como la duración del vídeo es de 1:37 min tenemos una duración de <b> 97 segundos </b>
- Por lo tanto <b> en cada segundo se han obtenido 30 frames </b>, esto quiere decir: 
    - Ejericio 1 comprede desde el frame 0 hasta el frame 960
    - Ejericio 2 comprede desde el frame 960 hasta el frame 1530
    - Ejericio 3 comprede desde el frame 1530 hasta el frame 2160
    - Ejericio 4 comprede desde el frame 2160 hasta el frame 2919
    
Para mostrar los ejemplos, se va a empezar cargando los ángulos.

<div class="alert alert-block alert-warning">
    <b>Nota: </b> En este cuaderno se mostrarán las distintas fases de actuación, aunque en las pruebas se han ejecutada todos y cada uno de los ejemplos, en este cuaderno solo me mostrarán algunos por no crear un cuaderno inmenso.
</div> 

<div class="alert alert-block alert-warning">
    <b>Nota: </b> Los vídeos utilizados para esta prueba han sido, Ejercicio1.mp4, Ejercicio2.mp4, Ejercicio3.mp4, Ejercicio4.mp4, VideoCompleto.mp4
</div>

In [59]:
import pickle
import numpy as np

def extrat_pickle(file):
    pos=[]
    with open("/.Pickle/Angulos_completos/"+file, "rb") as f:
        while True: 
            try:
                current_id=pickle.load(f)
                pos.append(current_id)
            except EOFError:
                break
    return pos

In [60]:
files_ang = ["ang_ejercicio1.pickle", "ang_ejercicio2.pickle", "ang_ejercicio3.pickle", 
         "ang_ejercicio4.pickle", "ang_ejerciciosCompletos1.pickle"]

for i,f in enumerate(files_ang,1):
    locals()["array_angulos" + str(i)] = extrat_pickle(f)

array_angulos_totales = array_angulos5

## Solución 1: Búsqueda de toda la secuencia en el vídeo completo

La primera línea de actuación será buscar la secuencia completa dentro del vídeo. En este apartado se comentarán algunos problemas que han surgido al realizar la búsqueda.

In [61]:
print("Longitud del array que contiene el Ejercicio1: ", len(array_angulos1))
print("Longitud del array que contiene el Ejercicio2: ", len(array_angulos2))
print("Longitud del array que contiene el Ejercicio3: ", len(array_angulos3))
print("Longitud del array que contiene el Ejercicio4: ", len(array_angulos4))
print("\nLongitud del array que contiene todos los ejercicios: ", len(array_angulos_totales))

Longitud del array que contiene el Ejercicio1:  969
Longitud del array que contiene el Ejercicio2:  624
Longitud del array que contiene el Ejercicio3:  626
Longitud del array que contiene el Ejercicio4:  439

Longitud del array que contiene todos los ejercicios:  2897


##### Búsqueda del primer ejercicio

In [10]:
from tslearn.metrics import dtw_subsequence_path

path, dist = dtw_subsequence_path(array_angulos1, array_angulos_totales)

path=np.array(path)
a_ast = path[0, 1]
b_ast = path[-1, 1]

print("El Ejercicio 1 comienza en el frame =",a_ast)
print("El Ejercicio 1 finaliza en el frame =",b_ast)

El Ejercicio 1 comienza en el frame = 1177
El Ejercicio 1 finaliza en el frame = 1177


<div class="alert alert-block alert-warning">
    <b>Nota: </b> 
Que el resultado sea una secuencia que empieza y acaba en el mismo punto quiere indicar que en dicho punto se encuentra algún valor nulo. Para subsanar este problema, se deberán eliminar todas las posiciones que contengan dichos valores. 
 </div> 

In [63]:
array_angulos_totales.pop(1177)

[nan,
 nan,
 181.80103391290598,
 195.42391451438812,
 263.0643529882258,
 255.47103523185842,
 259.52708916343613,
 270.0585239928155,
 87.44936103700151,
 92.55063916759345,
 195.60737339549468,
 177.72324341140262]

In [64]:
path, dist = dtw_subsequence_path(array_angulos1, array_angulos_totales)

path=np.array(path)
a_ast = path[0, 1]
b_ast = path[-1, 1]

print("El Ejercicio 1 comienza en el frame =",a_ast)
print("El Ejercicio 1 finaliza en el frame =",b_ast)

El Ejercicio 1 comienza en el frame = 944
El Ejercicio 1 finaliza en el frame = 1294


<div class="alert alert-block alert-danger">
Para comprobar porque arroja un resultado tan pequeño se ha probado a acotar el vídeo de referencia.
    
Si acotamos la búsqueda: 
- No tiene ningún sentido que encuentre una secuencia que comienza en el frame 716 y acaba en el 924 ya que al ser la secuencia de ejercicios completa, como mínimo deberia de abarcar unos 900 frames.

Si no la acotamos:
- Su camino óptimo es empezar en el frame 944 y acabar en el frame 1294, no tiene ningún sentido y empieza una vez se ha acabado el ejercicio.
    
En ambas ocasiones el problema es que detecta la secuencia demasiado tarde
</div>

Con un ejemplo vamos a dar una breve explicación de porque puede estar pasando esto:
- Patrón de referencia 
  [[10,4,45,6],[34,67,2,56],[34,56,78]]
- Secuencia en la que buscar 
  [[564,444,435,2346],[3344,6347,32,5634],[33244,5346,7438],
  [10,400,465,6],[3454,6547,52,556],<b>[60,40,80]</b>,
  [10234,3424,435,634],[3344,6347,342,5346],[3465,556,778]]
  
Como se puede obser el patrón de referencia es una matriz compuesta por 3 filas y la matriz en la que se pretende buscar dichos valores tiene un total de 9 filas, ambas tienen idéntico número de columnas.
Lo que hace el algoritmo de tslearn para la búsqueda de secuencias es mirar el patrón y buscar una similitud dentro de la cadena larga pero, como solo encuentra similitud con una parte del patrón, es esa la que devuelve.
Esto es lo que pasa en la búsqueda de secuencias de nuestro ejemplo, como no encuentra coincidencias bastante parecidas con el patrón, simplemente las descarta y por eso saca muchos menos frames de los esperados.

##### Búsqueda del segundo ejercicio

In [26]:
path, dist = dtw_subsequence_path(array_angulos2, array_angulos_totales)

path=np.array(path)
a_ast = path[0, 1]
b_ast = path[-1, 1]

print("El Ejercicio 2 comienza en el frame =",a_ast)
print("El Ejercicio 2 finaliza en el frame =",b_ast)

El Ejercicio 2 comienza en el frame = 1853
El Ejercicio 2 finaliza en el frame = 1856


<div class="alert alert-block alert-danger">
    En este caso la secuencia que encuentra es ridículamente pequeña y no se ubica dentro de las posibles soluciones ya que detecta la secuencia de inicio después de la secuencia de fin real. 
</div>

##### Búsqueda del tercer ejercicio

In [27]:
path, dist = dtw_subsequence_path(array_angulos3, array_angulos_totales)

path=np.array(path)
a_ast = path[0, 1]
b_ast = path[-1, 1]

print("El Ejercicio 3 comienza en el frame =",a_ast)
print("El Ejercicio 3 finaliza en el frame =",b_ast)

El Ejercicio 3 comienza en el frame = 357
El Ejercicio 3 finaliza en el frame = 643


<div class="alert alert-block alert-danger">
    Nuevamente encuentra una secuencia demasiado pequeña y en posiciones incorrectas
</div>

##### Búsqueda del cuarto ejercicio

In [28]:
path, dist = dtw_subsequence_path(array_angulos4, array_angulos_totales)

path=np.array(path)
a_ast = path[0, 1]
b_ast = path[-1, 1]

print("El Ejercicio 4 comienza en el frame =",a_ast)
print("El Ejercicio 4 finaliza en el frame =",b_ast)

El Ejercicio 4 comienza en el frame = 2610
El Ejercicio 4 finaliza en el frame = 2794


<div class="alert alert-block alert-danger">
    Nuevamente encuentra una secuencia demasiado pequeña y en posiciones incorrectas
</div>

## Solución 2: Buscar el inicio y el final

En este caso lo que se va a plantear es, mediante la búsqueda de secuencias, se sacará el principio y el fin de cada ejercicico corto. Una vez se tienen localizados esos valores, dentro del ejercicio largo se buscarán todas las posibles coincidencias y se irán almacenando por orden de llegada. Con las secuencias almacenada se procederá a hacer las siguientes comparaciones:
- De las secuencias de inicio se almacenará únicamente donde comienzan (a*)
- De las secuencias de final se almacenará únicamente donde finalizan (b*)
- Las comparaciones de secuencias se realizan entre aquellas que:
    - a* < b*
    - b* - a* > longitud de los frames * tasa de error
    
Ahora buscaremos y almacenaremos todas las secuencias de inicio y de fin dentro del array que contiene los todos los ejercicios

In [44]:
from scipy.signal import find_peaks

from tslearn import metrics
from tslearn.generators import random_walks
from tslearn.preprocessing import TimeSeriesScalerMeanVariance


def calcule_matrix(X,Y):
    """
    Función que calcula la matriz de costes locales o matriz de distancias.
    Este cálculo se ha realizao mediante la distancia euclidiana.
    
    Parámetros de entrada 
    ---------------------
    X : Secuencia corta con un ejercicio concreto.
    Y : Secuencia larga con múltiples ejercicios.
    
    Salida
    ------
    Matriz de costes locales o de distancias.
    """
    C=np.zeros((len(X),len(Y)))
    for i in range(len(X)):
        for j in range(len(Y)):
            x=np.array(X[i])
            y=np.array(Y[j])
            
            #Cálculo de la distancia euclídea entre dos secuencias.
            C[i][j] = np.linalg.norm(np.array(X[i])-np.array(Y[j])) #para secuencias de varias dimensiones

    return C

def calcule_matrixD(C,X,Y):
    """
    Función que calcula la matriz de costes acumulados.
    Este cálculo se realizará de la siguiente forma:
        1) La primera fila de D está inicializada como D(1,m):=C(1,m) para m € [1:M] siendo M=len(Y).
        2) La primera columna de D está inicializada como D(n,1):=∑k=1nC(k,1) para n € [1:N] siendo N=len(X).
        3) El resto de los valores se definen recursivamente a partir de la matriz de costes locales.
    
    Parámetros de entrada 
    ---------------------
    C : Matriz de costes locales o de distancias.
    X : Secuencia corta con un ejercicio concreto.
    Y : Secuencia larga con múltiples ejercicios.
    
    Salida
    ------
    Matriz de costes acumulados.
    """
    D=np.zeros((len(X),len(Y)))
    D[:,0]=np.cumsum(C[:,0]) #inicialización de la primera columna
    D[0,:]=C[0,:] #Inicialización de la primera fila 

    for i in range(1,len(X)):
        for j in range(1,len(Y)):
            D[i][j] = C[i][j]+min(D[i-1][j],D[i][j-1],D[i-1][j-1])
            
    return D

def calcule_path(C,sz):
    """
    Función que calcula las posibles rutas de deformación.
    Estas rutas serán calculadas mediante la función scipy.signal.find_peaks que localiza picos dentro
    de una señal. Con el argumento "distance" se especifica distancia horizontal mínima entre picos, cuanto 
    más restrictivo sea este parámetro, menor será el número de caminos localizados.
    
    Parámetros de entrada 
    ---------------------
    C : Matriz de costes locales o de distancias.
    sz : Parámetro que se utiliza para hacer más o menos restrictiva la búsqueda.
    
    Salida
    ------
    Posibles rutas de deformación.
    """
    #Calculo del coste de la función
    cost_func = C[-1, :]

    #Identificación de los posibles caminos
    potential_matches = find_peaks(-cost_func, distance=sz * 4.75, height=-50)[0]

    #Calculo de las rutas óptimas a partir de cada uno de los mínimos identificados
    paths = [metrics.subsequence_path(C, match) for match in potential_matches]
    
    return paths

In [45]:
# Las longitudes de inicio y final se han realizado a ojo simplemente para realizar las pruebas
inicio_ej1=array_angulos1[:100] ; fin_ej1=array_angulos1[-100:]
inicio_ej2=array_angulos1[:150] ; fin_ej2=array_angulos1[-150:]
inicio_ej3=array_angulos1[:200] ; fin_ej3=array_angulos1[-200:]
inicio_ej4=array_angulos1[:250] ; fin_ej4=array_angulos1[-250:]

inicios = [inicio_ej1,inicio_ej2,inicio_ej3,inicio_ej4] 
finales = [fin_ej1,fin_ej2,fin_ej3,fin_ej4]

supuestos_inicios=[]
supuestos_finales=[]

for i, ini in enumerate(inicios,1):
    C=calcule_matrix(ini, array_angulos_totales)
    paths=calcule_path(C,70)
    print("Número de posibles inicios para el ejercicio"+str(i)+": ",len(paths))
    #Por cada una se sacan los supuestos inicios
    for p in paths:
        p=np.array(p)
        a_ast = p[0, 1]
        supuestos_inicios.append(a_ast)
    locals()["supuestos_inicios" + str(i)]=supuestos_inicios
    
print("\n")    
for i, fin in enumerate(finales,1):
    C=calcule_matrix(fin, array_angulos_totales)
    paths=calcule_path(C,70)
    print("Número de posibles finales para el ejercicio"+str(i)+": ",len(paths))
    #Por cada una se sacan los supuestos finales
    for p in paths:
        p=np.array(p)
        b_ast = p[-1, 1]
        supuestos_finales.append(b_ast)
    locals()["supuestos_finales" + str(i)]=supuestos_finales


Número de posibles inicios para el ejercicio1:  6
Número de posibles inicios para el ejercicio2:  8
Número de posibles inicios para el ejercicio3:  5
Número de posibles inicios para el ejercicio4:  7


Número de posibles finales para el ejercicio1:  7
Número de posibles finales para el ejercicio2:  7
Número de posibles finales para el ejercicio3:  7
Número de posibles finales para el ejercicio4:  7


In [47]:
print("Supuestos inicios para el ejercicio 1", supuestos_inicios1)
print("Supuestos finales para el ejercicio 1", supuestos_finales1)

print("\nSupuestos inicios para el ejercicio 2", supuestos_inicios2)
print("Supuestos finales para el ejercicio 2", supuestos_finales2)

print("\nSupuestos inicios para el ejercicio 3", supuestos_inicios3)
print("Supuestos finales para el ejercicio 3", supuestos_finales3)

print("\nSupuestos inicios para el ejercicio 4", supuestos_inicios4)
print("Supuestos finales para el ejercicio 4", supuestos_inicios4)

Supuestos inicios para el ejercicio 1 [245, 697, 1253, 1614, 2170, 2528, 0, 287, 697, 1153, 1469, 1801, 2192, 2670, 104, 431, 1001, 2287, 2841, 0, 431, 838, 1436, 1755, 2213, 2528]
Supuestos finales para el ejercicio 1 [122, 633, 974, 1383, 1767, 2204, 2762, 122, 633, 974, 1383, 1767, 2204, 2762, 122, 633, 974, 1383, 1767, 2204, 2762, 122, 633, 974, 1383, 1767, 2204, 2762]

Supuestos inicios para el ejercicio 2 [245, 697, 1253, 1614, 2170, 2528, 0, 287, 697, 1153, 1469, 1801, 2192, 2670, 104, 431, 1001, 2287, 2841, 0, 431, 838, 1436, 1755, 2213, 2528]
Supuestos finales para el ejercicio 2 [122, 633, 974, 1383, 1767, 2204, 2762, 122, 633, 974, 1383, 1767, 2204, 2762, 122, 633, 974, 1383, 1767, 2204, 2762, 122, 633, 974, 1383, 1767, 2204, 2762]

Supuestos inicios para el ejercicio 3 [245, 697, 1253, 1614, 2170, 2528, 0, 287, 697, 1153, 1469, 1801, 2192, 2670, 104, 431, 1001, 2287, 2841, 0, 431, 838, 1436, 1755, 2213, 2528]
Supuestos finales para el ejercicio 3 [122, 633, 974, 1383, 1767,

Como se puede observar, en todos los inicios y finales hay valores muy dispersos, por lo que esto generará secuencias muy distintas.
Para mostrar un ejemplo de una posible clasificaión se hará una única prueba con las secuencias de inicio y fin obtenidas a partir del cuarto ejercicio

In [50]:
from tslearn.metrics import dtw_subsequence_path
from dtaidistance import dtw_ndim
from pydtw import dtw2d
from tslearn.metrics import dtw_path_from_metric
from tslearn.metrics import dtw, dtw_path
from tslearn.metrics import soft_dtw


def classify(array_angulos4,array_angulos_totales):
    lon_ej1=len(array_angulos4)
    error=lon_ej1*0.2
    seq_ej1=[]
    rate_error=20

    for a in supuestos_inicios4:
        for b in supuestos_inicios4:
            diferencia=b-a
            if a < b and diferencia < lon_ej1+error and diferencia > lon_ej1-error:
                print("Inicio: ",a)
                print("Final: ",b)

                serie1=np.array(array_angulos4)
                serie2=np.array(array_angulos_totales[a:b+1])

                path, dist = dtw_subsequence_path(serie1, serie2)
                print("Distancia 1: ",dist)

                distance = dtw_ndim.distance(serie1,serie2)
                print("Distancia 2: ",distance)

                euclidean_distance = dtw_ndim.ub_euclidean(serie1,serie2)
                print("Euclidean_distance: ",euclidean_distance)

                cost_matrix, cost, alignmend_a, alignmend_b = dtw2d(serie1,serie2)
                print('Coste 1: ',cost)

                optimal_path, cost = dtw_path_from_metric(serie1, serie2)
                print("Coste 2: ",cost)

                cost = dtw(serie1, serie2, global_constraint="sakoe_chiba", sakoe_chiba_radius=0.5)
                print("Coste 3: ",cost)

                cost = dtw(serie1, serie2, global_constraint="itakura", itakura_max_slope=2.)
                print("Coste 4: ",cost)

                soft_dtw_score = soft_dtw(serie1, serie2, gamma=1.5)
                print("Soft: ",soft_dtw_score)
                
classify(array_angulos4,array_angulos_totales)            

Inicio:  245
Final:  697
Distancia 1:  3611.004558605163
Distancia 2:  4036.70547511661
Euclidean_distance:  5353.2085699453655
Coste 1:  20645816.113429584
Coste 2:  79930.91400297038
Coste 3:  4132.662190832851
Coste 4:  4056.645992273918
Soft:  16294991.092667092
Inicio:  245
Final:  697
Distancia 1:  3611.004558605163
Distancia 2:  4036.70547511661
Euclidean_distance:  5353.2085699453655
Coste 1:  20645816.113429584
Coste 2:  79930.91400297038
Coste 3:  4132.662190832851
Coste 4:  4056.645992273918
Soft:  16294991.092667092
Inicio:  697
Final:  1153
Distancia 1:  3580.3459139961515
Distancia 2:  3994.1726419403603
Euclidean_distance:  5337.520895795244
Coste 1:  inf
Coste 2:  79794.74652683332
Coste 3:  4115.445992279476
Coste 4:  4062.7847069342447
Soft:  15953415.093624836
Inicio:  1253
Final:  1614
Distancia 1:  3682.9861116006728
Distancia 2:  3891.7726587944167
Euclidean_distance:  5023.284512967302
Coste 1:  17893175.037571516
Coste 2:  73255.699079317
Coste 3:  3941.38196412

<div class="alert alert-block alert-danger">

<b> PROBLEMAS </b>
- La primera coincidencia encuentra más o menos la secuencia, es decir, ni siqueiera encuentra una secuencia fiable.
- Localiza DEMASIADAS secuencias.
- La distancia DTW es similar entre todas, de hecho, la secuencia correcta tiene una distancia DTW mayor que otras secuencias incorrectas.
    
    
En principio podría pensarse que esto se puede deber a que el inicio y el final del Ejercicio 4 puede asemejarse bastante a otras partes o movimientos de otros ejercicios, pero tras probar con todos y cada uno de los ejercicios se ha demostrado que los resultados son muy parecidos
    
</div>

## Solución 3: Buscar una secuencia intermedia 

Para esta posible solución se va a obtener una secuencia intermedia del Ejercicio 1. Esta secuencia se obtiene a mano tras analizar la longitud de la secuencia respectiva a al ejercicio mencionado. 

In [66]:
middel_ej1 = array_angulos1[387:542]

In [67]:
from tslearn.metrics import dtw_subsequence_path

path, dist = dtw_subsequence_path(middel_ej1, array_angulos_totales)
#print(path)
#print(dist)

path=np.array(path)
a_ast = path[0, 1]
b_ast = path[-1, 1]

print("El Ejercicio 1 comienza en el frame =",a_ast)
print("El Ejercicio 1 finaliza en el frame =",b_ast)
#print("La subsecuencia que debería encontrar es: ",array_angulos1)
#array_angulos_total[a_ast:b_ast+1]

El Ejercicio 1 comienza en el frame = 2520
El Ejercicio 1 finaliza en el frame = 2574


<div class="alert alert-block alert-danger">
    Nuevamente, se puede observar que no encuentra correctamente el ejercicio.
</div>

Una posible opción seria proceder como en el caso anterior. Buscar todas las posibles secuencias y después clasificarlas. 

Como se ha comprobado que la forma de clasificarlas no es idónea para este estudio, este apartado simplemente se deja planteado

In [58]:
C=calcule_matrix(np.array(middel_ej1), np.array(array_angulos_totales))
print("\nMatriz de distancias o matriz de costes locales con sqrt(x[i] - y[j])^2:\n ",C) 

cost_func = C[-1, :]

sz=90
potential_matches = find_peaks(-cost_func, distance=sz * 4.75, height=-50)[0]

paths = [metrics.subsequence_path(C, match) for match in potential_matches]
print("\n Número de caminos óptimos: ",len(paths))

supuestos_finales=[]
supuestos_inicios=[]
for p in paths:
    p=np.array(p)
    a_ast = p[0, 1]
    b_ast = p[-1, 1]
    supuestos_inicios.append(b_ast)
    supuestos_finales.append(b_ast)
#supuestos_finales


Matriz de distancias o matriz de costes locales con sqrt(x[i] - y[j])^2:
  [[239.8002441  222.44576493 238.121359   ... 137.03028567 233.33104238
  139.83000002]
 [249.53563278 146.51855963 244.41735526 ... 146.07652176 244.76971027
  149.29979946]
 [205.47466906 235.07217244 205.14447666 ... 157.55064019 199.72730947
  164.91090677]
 ...
 [194.98593599  42.62383054 188.89073083 ... 199.23474985 191.36985163
  209.79279671]
 [206.12860504 267.30499585 207.12010949 ... 196.61849588 202.41559284
  205.20614482]
 [243.32230263 221.51221214 241.55175408 ... 137.54519525 236.83244581
  139.69336705]]

 Número de caminos óptimos:  6


<div class="alert alert-block alert-danger">

Se podría pensar que como los ejercicios intermedios tienden a ser más distintos entre ellos, por ejemplo, los ejercicios casi siempre empieza y termina en un estado de reposo, analizando las secuencias intermedias se iban a obtener mejores resultados, pero la realidad ha sido la misma que en el caso anterior.

- Sigue sacando un montón de posibles rutas de deformación.
- Los valores de peso y distancia son demasiado parecidos para poder acotar las secuencias.
- Al acotar con valores demasiado ambiciosos acaba eliminando la secuencia correcta.
    
</div>

<div class="alert alert-block alert-warning">
<b> CONCLUSIÓN </b>
Ninguna de las técnicas mostradas en este cuaderno es válida para la búsqueda de secuencias compuestas por posiciones de esqueleto
</div>
